### PR TITLE
[Feat] Support MTP to running in full graph mode

### PR DESCRIPTION
**Pull request by anon189Ty**
### What this PR does / why we need it?
Currently, the MTP model still runs in eager in full graph mode. This PR adapts the MTP with the full graph capture and execution. When the graph mode is set to "FULL_DECODE_ONLY", the MTP will run in full-graph to improve the performance.

The change in both disable_padded_drafter_batch is True and False case include:

1. Add _mtp_graph_params in acl_graph.py to isolate the data of main model and the data of MTP.
2. Padding some metadata in mla_v1.py when in fullgraph mode.
3. Fixed the essential data address that will be used in model.forward.
4. Adapted according to the aclgraph capture framwork:
    1). Rebuild MTP model with ACLGraphWrapper.
    2). Add common attn metadata when start capture in MTP dummy_run.
    3). Add common attn metadata update in MTP.
    4). Addapted data update when num_speculative_tokens > 1.
5. Add a patch of MTP to adapt vllm v0.11.0.

Existing Issues:
  1. When disable_padded_drafter_batch=True and running in FullGraph mode, the data of the first-round requests in MTP is abnormal. We need to identify the cause subsequently.
  2. When disable_padded_drafter_batch=False and running in FullGraph mode, the acceptance rate of the second and third tokens will decrease (For example, if we set the num_speculative_tokens=3, the acceptance rate of first token is 90%, the second is only 50% lower than 60%, the third is only 20% lower than 30%). The reason is that the data processed after the model runs does not match. This is a problem from another PR. It works fine in eager and PIECEWISE mode, but has problem in FullGraph mode. Once we have a solution, we will submit a bugfix.

### How was this patch tested?

- vLLM version: v0.11.0
- vLLM main: https://github.com/vllm-project/vllm/commit/2918c1b49c88c29783c86f78d2c4221cb9622379

### CODE REVIEW

File: vllm_ascend/ascend_forward_context.py [hunk 1]
```diff
@@ -157,6 +158,7 @@ def set_ascend_forward_context(
         forward_context.prefetch_mlp_enabled = prefetch_mlp_enabled
         forward_context.model_instance = model_instance
         forward_context.weight_prefetch_method = weight_prefetch_method
+        forward_context.is_mtp_model = is_mtp_model
```
Comment (yiz-liu) [inline vllm_ascend/ascend_forward_context.py:162]: We should try to remove this flag in the future.

File: vllm_ascend/attention/mla_v1.py [hunk 2]
```diff
@@ -1447,8 +1573,11 @@ def _forward_decode(
             "actual_seq_lengths": actual_seq_lengths,
             "actual_seq_lengths_kv": decode_meta.seq_lens_list,
         }
-        graph_params = get_graph_params()
         forward_context: ForwardContext = get_forward_context()
+        if forward_context.is_mtp_model:
+            graph_params = get_mtp_graph_params()
+        else:
+            graph_params = get_graph_params()
```
Comment (yiz-liu) [inline vllm_ascend/attention/mla_v1.py:1403]: Graph params is also deviating from its initial design and having more and more if-else.

File: vllm_ascend/compilation/acl_graph.py [hunk 3]
```diff
@@ -278,6 +282,12 @@ def update_mla_attn_params(update_stream, forward_context, runtime_shape,
                     spec_multiple * (i + 1)
                     for i in range(runtime_shape // spec_multiple)
                 ]
+            elif forward_context.is_mtp_model:
+                actual_seq_lengths = forward_context.attn_metadata[
+                    key].decode.actual_seq_lengths_q
+                block_table = forward_context.attn_metadata[key].decode.block_table
+                seq_lens_list = seq_lens_list + [0] * (len(actual_seq_lengths) -
+                                                       len(seq_lens_list))
```
Comment (yiz-liu) [inline vllm_ascend/compilation/acl_graph.py]: If we can delete this, then `is_mtp_model` is no longer necessary right?

File: vllm_ascend/patch/worker/patch_deepseek_mtp.py [hunk 4]
```diff
@@ -51,4 +67,161 @@ def predictor_init(self, vllm_config: VllmConfig, prefix: str) -> None:
...
+
+    hidden_states, residual = self.mtp_block(positions=positions,
+                                             hidden_states=hidden_states,
+                                             residual=None)
+    hidden_states = residual + hidden_states
+    return hidden_states
+
+
+@support_torch_compile
+class CustomDeepSeekMTP(nn.Module, SupportsPP):
```
Comment (yiz-liu) [inline vllm_ascend/patch/worker/patch_deepseek_mtp.py]: Maybe try slim this.

File: vllm_ascend/spec_decode/mtp_proposer.py [hunk 5]
```diff
@@ -751,12 +874,19 @@ def _propose(
...
+                    attn_metadata_i.decode.seq_lens_list = decode_seq_lens_list + [
+                        0
+                    ] * (graph_pad_size - len(decode_seq_lens_list))
                 attn_metadata_i.decode.input_positions = self.positions[:
                                                                         num_input_tokens]
                 attn_metadata_i.decode.max_seq_lens += 1
                 attn_metadata_i.decode.max_seq_lens = min(
                     attn_metadata_i.decode.max_seq_lens,
                     self.runner.model_config.max_model_len)
+            torch.npu.synchronize()
```
Comment (yiz-liu) [inline vllm_ascend/spec_decode/mtp_proposer.py:889]: Remove this line after #4233 being merged.

File: vllm_ascend/attention/mla_v1.py [hunk 6]
```diff
@@ -343,6 +345,74 @@ def reorder_batch(self, input_batch: "InputBatch",
...
+            assert len(interpolated) == len(padding_seq_len_q)
+            actual_seq_lengths_q = actual_seq_lengths_q + interpolated
+        else:
+            actual_seq_lengths_q = actual_seq_lengths_q + common_attn_metadata.actual_seq_lengths_q[
+                num_reqs:num_reqs + num_reqs_pad_size]
+
+        return actual_seq_lengths_q
+
+    def pad_actual_seq_len_q_mtp_disable_pad(self, num_reqs_pad_size, num_reqs,
+                                             actual_seq_lengths_q):
```
Comment (zouyida2052) [inline vllm_ascend/attention/mla_v1.py]: In this case, when we enable num_mtp_tokens=3, the acceptance rate in positions 2 and 3 is 0 in first request.

File: vllm_ascend/attention/mla_v1.py [hunk 7]
```diff
@@ -343,6 +345,74 @@ def reorder_batch(self, input_batch: "InputBatch",
         # better way of doing this
         return modified_batch
 
+    def pad_actual_seq_len_q_mtp_enable_pad(self, num_reqs_pad_size, num_reqs,
+                                            actual_seq_lengths_q,
+                                            common_attn_metadata):
```
Comment (zouyida2052) [inline vllm_ascend/attention/mla_v1.py]: In this case, mean acceptance length seems to be a little lower than piecewise mode or eager mode.

File: vllm_ascend/attention/mla_v1.py [hunk 8]
```diff
@@ -343,6 +345,74 @@ def reorder_batch(self, input_batch: "InputBatch",
...
+            assert interpolated[-1] == end_val
+            assert len(interpolated) == len(padding_seq_len_q)
+            actual_seq_lengths_q = actual_seq_lengths_q + interpolated
+        else:
+            actual_seq_lengths_q = actual_seq_lengths_q + common_attn_metadata.actual_seq_lengths_q[
+                num_reqs:num_reqs + num_reqs_pad_size]
+
+        return actual_seq_lengths_q
+
+    def pad_actual_seq_len_q_mtp_disable_pad(self, num_reqs_pad_size, num_reqs,
```
Comment (whx-sjtu) [inline vllm_ascend/attention/mla_v1.py:385]: The function name `pad_actual_seq_len_q_mtp_disable_pad` looks confusing to me. If the function is used to pad `actual_seq_len_q`, why the name ended with diable_pad?

File: vllm_ascend/attention/mla_v1.py [hunk 9]
```diff
@@ -343,6 +345,74 @@ def reorder_batch(self, input_batch: "InputBatch",
         # better way of doing this
         return modified_batch
 
+    def pad_actual_seq_len_q_mtp_enable_pad(self, num_reqs_pad_size, num_reqs,
```
Comment (whx-sjtu) [inline vllm_ascend/attention/mla_v1.py:342]: ditto

File: vllm_ascend/attention/mla_v1.py (hunk 8 above)
Comment (anon189Ty) [reply vllm_ascend/attention/mla_v1.py:385]: The "disable_pad" in function name means "disable_padded_drafter_batch=True", but it's too long. Similarly, the "enable_pad" means "disable_padded_drafter_batch=False".
